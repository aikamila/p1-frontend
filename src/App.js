import logo from './logo.svg';
import './App.css';
import { BrowserRouter as Router, Route, Switch} from 'react-router-dom';
import StartPage from './pages/StartPage';
import PrivateRoute from './utils/PrivateRoute';
import PublicRoute from './utils/PublicRoute';

import AuthPage from './pages/AuthPage';
import HomePage from './pages/HomePage';
import InitialVerificationPage from './pages/InitialVerificationPage';
import UnsuccessfulInitialVerificationPage from './pages/UnsuccessfulInitialVerificationPage';

function App() {
  return (
     <div className="App">

          <PublicRoute path="/" exact>
            <StartPage/>
          </PublicRoute>
          <PublicRoute path="/auth">
            <AuthPage/>
          </PublicRoute>
          <PrivateRoute path="/home"> 
            <HomePage/>
          </PrivateRoute>
            <Switch>
              <Route path="/verification/unsuccessful" component={UnsuccessfulInitialVerificationPage}/>
              <Route path="/verification/:id/:token" component={InitialVerificationPage} />
            </Switch>

     </div>

  );
}

export default App;
