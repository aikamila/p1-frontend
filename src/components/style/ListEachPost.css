.list-each-post__container {
  width: 90%;
  /* height: 15em; */
  background-color: white;
  border: 1px solid lightgray;
  border-radius: 3px;
  margin-top: 3%;
  margin-bottom: 3%;
  padding: 3% 5%;
}

.list-each-post__username {
  color: black;
  font-weight: 500;
  text-decoration: none;
}

.list-each-post__content {
  margin-bottom: 0%;
}

.list-each-post__see_more {
  color: var(--nice-blue);
  text-decoration: none;
}

.list-each-post__update_link {
  color: var(--nice-blue);
  text-decoration: none;
}

.list-each-post__time {
  color: gray;
  font-size: 0.8em;
}

.list-each-post__comment_count {
  font-size: 0.8em;
  color: gray;
}

.list-each-post__content {
  overflow: hidden;
}


