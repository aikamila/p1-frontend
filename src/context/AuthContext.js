import { createContext, useState, useEffect, useCallback } from "react";
import { useHistory, useLocation } from "react-router-dom";
import { getItemAndCheckExpiry, setItemwith8MonthExpiry } from "../utils/LocalStorage";
import { calculateTime } from "../utils/Timing";
import { decodeUser } from "../utils/JWTDecode";

const AuthContext = createContext()

export default AuthContext;

export const AuthProvider = ({children}) => {

    const history = useHistory()

    let location = useLocation()

    const [authTokens, setAuthTokens] = useState(null)
    const [userId, setUserId] = useState(null)

    const [loginEmail, setLoginEmail] = useState('')
    const [loginPassword, setLoginPassword] = useState('')

    // const [pwResetEmail, setPwResetEmail] = useState('')
    // const [pwResetPassword, setPwResetPassword] = useState('')
    // const [pwResetPasswordSecond, setPwResetPasswordSecond] = useState('')

    const [loginError, setLoginError] = useState(null) // invalid credentials
 
    const [authServerError, setAuthServerError] = useState(null) // server error while logging in
    const [authSignUpServerError, setAuthSignUpServerError] = useState(null) // server error while signing up
    const [unexpectedLogoutError, setUnexpectedLogoutError] = useState('Something went wrong and you were logged out.')
    const [initialVerificationError, setInitialVerificationError] = useState(null)
    const [initialVerificationServerError, setInitialVerificationServerError] = useState(null)
    const [initialVerificationSuccessful, setInitialVerificationSuccessful] = useState(true)

    const [loginButtonDisabled, setLoginButtonDisabled] = useState(false)

    const BACKEND_DOMAIN = 'https://arcane-spire-03245.herokuapp.com/'
    const TOKENS_OBTAIN_URL = BACKEND_DOMAIN + 'api/user/token/'
    const TOKENS_REFRESH_URL = BACKEND_DOMAIN + 'api/user/token/refresh/'
    const TOKENS_BLACKLIST_URL = BACKEND_DOMAIN + 'api/user/token/blacklist/'
    const INITIAL_EMAIL_VERIFICATION = BACKEND_DOMAIN + 'api/user/token/verification/'

    const login = useCallback(async (e) => {
        e.preventDefault();
        setLoginButtonDisabled(true)
        try{
            const response = await fetch(TOKENS_OBTAIN_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'email': loginEmail, 'password': loginPassword})
            })
            const data = await response.json()
            if (response.status === 401) {
                throw Error('Invalid credentials')
            }else if (response.status === 200){
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setLoginPassword('')
                history.push('/home')
            }else if(loginEmail.length === 0 || loginPassword.length === 0){
                throw Error('Invalid credentials')
            }else{
                throw Error('Something else happened')
            }
        }catch (err){
            switch (err.message){
                case 'Invalid credentials':
                    setLoginError(err.message)
                    setLoginPassword('')
                    break;
                default:
                    setAuthServerError('Upps... unexpected problem with the server occured. Try to refresh the site and log in again :)')
            }
        }
        setLoginButtonDisabled(false)
    }, 
    [TOKENS_OBTAIN_URL, loginEmail, loginPassword, history]
    )

    const logout = useCallback(async () => {
        try{
            await fetch(TOKENS_BLACKLIST_URL,{
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'refresh': authTokens.refresh})
            })}catch(err){
                // doing nothing if the blacklisting of tokens goes wrong on the backend
            }
        const tokens = localStorage.getItem('authTokens')
        if (tokens){
            localStorage.removeItem('authTokens')
        }
        setAuthTokens(null)
        setUserId(null)
        history.push('/')
    }, [TOKENS_BLACKLIST_URL])

    const updateToken = useCallback(async () => {
        try{
            const tokens = getItemAndCheckExpiry('authTokens')
            // we get them from the local storage bc when the user comes back, they don't have the hook set (so we would send null :( )
            const response = await fetch(TOKENS_REFRESH_URL ,{
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({'refresh': tokens.refresh})
            })
            const data = await response.json()
            if (response.status !== 200) {
                throw Error('Invalid credentials')
            }else{
                setItemwith8MonthExpiry('authTokens', data)
                setAuthTokens(data)
                setUserId(decodeUser(data))
            }
        }catch(err){
            logout(); 
            setUnexpectedLogoutError('Something went wrong and you were logged out.')
            setTimeout(() => {
                setUnexpectedLogoutError(null)
            }, 5000)
        }
    }, [TOKENS_REFRESH_URL, logout]
    )

    const verifyEmail = async (id, token) => {
        try{
            const response = await fetch(INITIAL_EMAIL_VERIFICATION, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({'id': id, 'token': token})
                })
            if(response.status === 401){
                throw Error('Invalid link')
            }else if(response.status !== 200){
                throw Error("Something else happened")
            }
            const data = await response.json()
            if(response.status===200){
                await logout(); // if there is a user that is logged in, their token should be blacklisted
                setAuthTokens(data);
                setUserId(decodeUser(data))
                setItemwith8MonthExpiry('authTokens', data)
                setInitialVerificationSuccessful(true)
                setTimeout(() => {
                    setInitialVerificationSuccessful(false)
                }, 5000)
                history.push('/home')
            }
            }catch(err){
                if(err.message === 'Invalid link'){
                    setInitialVerificationError('The link is invalid. If you have already used it, just log into your account.')
                }else(
                    setInitialVerificationServerError('Something went wrong. Please use the link once again.')
                )
                history.push("/verification/unsuccessful")
            }
        
    }

    useEffect(() => {
        const fourMinutes = calculateTime(0, 4, 0)
        let interval = setInterval(() => {
            if (authTokens){
                updateToken()
            }
        }, fourMinutes)
        return () => clearInterval(interval)
    }, [authTokens, updateToken])

    useEffect(() => {
        const updateTokenAtTheBeginning = async () => {
            const tokens = getItemAndCheckExpiry('authTokens')
            if(tokens !== null){
                await updateToken();
            }
        }   
        updateTokenAtTheBeginning()
    }, [updateToken])

    useEffect(() => {
        setLoginPassword('')
        setLoginError(null)
        setAuthSignUpServerError(null)
        setAuthServerError(null)
    }, [location])

    let contextData = {
        login: login,
        loginEmail: loginEmail,
        setLoginEmail: setLoginEmail,
        loginPassword: loginPassword,
        setLoginPassword: setLoginPassword,
        loginError: loginError,
        authTokens: authTokens, 
        logout: logout,
        authServerError: authServerError,
        unexpectedLogoutError: unexpectedLogoutError,
        verifyEmail: verifyEmail,
        initialVerificationError: initialVerificationError,
        initialVerificationServerError: initialVerificationServerError,
        initialVerificationSuccessful: initialVerificationSuccessful,
        setAuthSignUpServerError: setAuthSignUpServerError,
        authSignUpServerError: authSignUpServerError,
        BACKEND_DOMAIN: BACKEND_DOMAIN,
        setLoginError: setLoginError,
        userId: userId,
        updateToken: updateToken,
        loginButtonDisabled: loginButtonDisabled,
    }

    return(
        <AuthContext.Provider value={contextData}>
            {children}
        </AuthContext.Provider>
    )}