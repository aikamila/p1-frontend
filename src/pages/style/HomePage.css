:root {
  --nice-blue: rgb(54, 155, 250);
  --background-gray: rgb(250, 250, 250);
  --disabled-blue: rgb(157, 200, 240);
  --fonts: "Raleway", "Segoe UI", "Roboto", "Oxygen", "Ubuntu", "Cantarell",
    "Fira Sans", "Droid Sans", "Helvetica Neue", sans-serif;
}

.home-page__email_verified_notification {
  margin: 0%;
  text-align: center;
  padding-bottom: 4px;
  padding-top: 3px;
  background-color: var(--background-gray);
}
