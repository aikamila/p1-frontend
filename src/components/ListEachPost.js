import React, { useContext } from 'react'
import AuthContext from '../context/AuthContext'
import { Link } from 'react-router-dom'
import './style/ListEachPost.css'

const ListEachPost = ({id, text, user, time, username, engagementRate}) => {
  const {userId} = useContext(AuthContext)
  return (
    <article className='list-each-post__container'>
      <Link className="list-each-post__username" to={`/home/account/${user}`}>{username}</Link><br></br>
      <span className='list-each-post__time'>{time}</span>
      {text.length < 121 ? <p className='list-each-post__content'>{text}</p> : <p className='list-each-post__content'>{`${text.slice(0,120)}...`}</p>}
      <Link className='list-each-post__see_more' aria-label='see more about this post on another page' to={`/home/post/${id}`}>See more...</Link><br></br>
      <span className='list-each-post__comment_count'>{engagementRate} comments</span><br></br>
      {userId === user && <Link className='list-each-post__update_link' aria-label='edit this post' to={`/home/post/${id}/update`}>Edit</Link>}
    </article>
  )
}

export default ListEachPost