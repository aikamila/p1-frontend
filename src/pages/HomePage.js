import React,{useContext} from 'react'
import AuthContext from '../context/AuthContext'
import { Switch, Route } from 'react-router-dom'
import ListPosts from '../components/ListPosts'
import PostDetails from '../components/PostDetails'
import UpdatePost from '../components/UpdatePost'
import AddPost from '../components/AddPost'
import UserAccount from '../components/UserAccount'
import './style/HomePage.css'


const HomePage = () => {
    let {initialVerificationSuccessful} = useContext(AuthContext)

    return (
        <>
            {initialVerificationSuccessful && <p role="alert">Your email was successfully verified!</p>}
            <Switch>
                <Route path="/home/account/:id" component={UserAccount} />
                <Route path="/home/post/:postId/update" component={UpdatePost}/>
                <Route path="/home/post/add" component={AddPost} />
                <Route path="/home/post/:postId" component={PostDetails}/>
                <Route path="/home">
                    <ListPosts />
                </Route>
            </Switch>
        </>
    )
}

export default HomePage

