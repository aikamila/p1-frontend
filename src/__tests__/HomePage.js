import { render, screen, waitFor } from '@testing-library/react'
import App from '../App'
import { AuthProvider } from '../context/AuthContext'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import userEvent from '@testing-library/user-event'
import AuthContext from '../context/AuthContext'
import HomePage from '../pages/HomePage'
import { server } from '../mocks/server'
import { rest } from 'msw'
import ListEachPost from '../components/ListEachPost'

afterEach(() => {
    localStorage.removeItem("authTokens")
})

jest.setTimeout(7000)

const renderHomePage = (history) => {
    return render(
        <Router history={history} >
            <AuthProvider>
                <App />
            </AuthProvider>
        </Router>
    )
}

const renderOnlyHomePage = () => {
    const history = createMemoryHistory();
    history.push("/home")
    const authTokens = {refresh: 'valid-refresh', access: 'valid-access'}
    return render(
    <Router history={history}>
        <AuthContext.Provider value={{logout: null, initialVerificationSuccessful: false, authTokens: authTokens,
        userId: 1, updateToken: null}}>
            <HomePage />
        </AuthContext.Provider>
    </Router>
    )
}

test("user is able to log out - when no problems occur", async () => {
    const history = createMemoryHistory()
    history.push("/auth")
    renderHomePage(history)
    userEvent.type(screen.getByRole("textbox", {name: /email/i}), "valid@mail.com")
    userEvent.type(screen.getByRole("textbox", {name: /password/i}), "Password*")
    userEvent.click(screen.getByRole("button", {name: /log in/i}))
    const button = await screen.findByRole("button", {name: /log out/i})
    userEvent.click(button)
    await waitFor(() => expect(localStorage.getItem("authTokens")).toBeNull)
    await waitFor(() => expect(history.location.pathname).toBe("/"))
    expect(screen.queryByText(/you were logged out/i)).not.toBeInTheDocument()
})

test("user is able to log out - when their refresh token is invalid", async () => {
    const history = createMemoryHistory()
    history.push("/auth")
    renderHomePage(history)
    userEvent.type(screen.getByRole("textbox", {name: /email/i}), "invalidtoken@mail.com")
    userEvent.type(screen.getByRole("textbox", {name: /password/i}), "Password*")
    userEvent.click(screen.getByRole("button", {name: /log in/i}))
    const button = await screen.findByRole("button", {name: /log out/i})
    userEvent.click(button)
    await waitFor(() => expect(localStorage.getItem("authTokens")).toBeNull)
    await waitFor(() => expect(history.location.pathname).toBe("/"))
    expect(screen.queryByText(/you were logged out/i)).not.toBeInTheDocument()
})

test("user is able to log out - when there is a server error", async () => {
    const history = createMemoryHistory()
    history.push("/auth")
    renderHomePage(history)
    userEvent.type(screen.getByRole("textbox", {name: /email/i}), "servererrortoken@mail.com")
    userEvent.type(screen.getByRole("textbox", {name: /password/i}), "Password*")
    userEvent.click(screen.getByRole("button", {name: /log in/i}))
    const button = await screen.findByRole("button", {name: /log out/i})
    userEvent.click(button)
    await waitFor(() => expect(localStorage.getItem("authTokens")).toBeNull)
    await waitFor(() => expect(history.location.pathname).toBe("/"))
    expect(screen.queryByText(/you were logged out/i)).not.toBeInTheDocument()
})

test("all posts are displayed to the user", async() => {
    // posts should be displayed on the home page, after the user logs in and all data is loaded
    renderOnlyHomePage()
    expect(screen.getByRole("img", {name: /wait a second/i})).toBeInTheDocument()
    expect(await screen.findByText(/Is anyone creating a Java app/i)).toBeInTheDocument()
    expect(screen.getByText(/anyone interested in helping me with a django app/i)).toBeInTheDocument()
    expect(screen.getByText('user1')).toBeInTheDocument()
    expect(screen.getByText('user2')).toBeInTheDocument()
    expect(screen.getByRole("link", {name:'user1'})).toHaveAttribute("href", "/home/account/1")
    expect(screen.getByRole("link", {name:'user2'})).toHaveAttribute("href", "/home/account/2")
    expect(screen.getByText('1 day ago')).toBeInTheDocument()
    expect(screen.getByText('2 days ago')).toBeInTheDocument()
    expect(screen.getAllByRole('article')).toHaveLength(2)   
    expect(screen.queryByRole("img", {name: /wait a second/i})).not.toBeInTheDocument()
    expect(screen.queryByText(/7 Comment\(s\)/i)).toBeInTheDocument()
    expect(screen.queryByText(/3 Comment\(s\)/i)).toBeInTheDocument()
    expect(screen.getAllByRole("link", {name: /see more/i})).toHaveLength(2)
    expect(screen.getByRole("region", {name: /recent posts/i})).toBeInTheDocument()
    expect(screen.getByRole("link", {name: /edit/i})).toBeInTheDocument()
    expect(screen.getByRole("link", {name: /edit/i})).toHaveAttribute("href", "/home/post/1/update")
})

test("server error while loading posts is handled gracefully", async() => {
    server.use(
        rest.get('https://arcane-spire-03245.herokuapp.com/api/services/posts/', (req, res, ctx) => {
          return res.once(
              ctx.status(404)
          )
        }),
    )
    renderOnlyHomePage()
    expect(screen.getByRole("img", {name: /wait a second/i})).toBeInTheDocument()
    expect(await screen.findByText(/We're not able to load the resources from the server/i)).toBeInTheDocument()
    expect(screen.queryByRole("img", {name: /wait a second/i})).not.toBeInTheDocument()
})

test("authorization error while loading posts is handled gracefully", async() => {
    const history = createMemoryHistory();
    history.push("/home")
    const authTokens = {refresh: 'valid-refresh', access: 'invalid-access'}
    const tokenUpdate = jest.fn()
         render(
        <Router history={history}>
            <AuthContext.Provider value={{logout: null, initialVerificationSuccessful: false, authTokens: authTokens,
            userId: 1, updateToken: tokenUpdate}}>
                <HomePage />
            </AuthContext.Provider>
        </Router>
        )
    
    expect(screen.getByRole("img", {name: /wait a second/i})).toBeInTheDocument()
    await waitFor(() => expect(tokenUpdate.mock.calls.length).toBe(1))
    // updateToken will receive a 401 response as well and will log the user out + the user will be informed
})

test("each post of a logged-in user is rendered properly", () => {
    // the post is rendered correctly and links redirect to valid urls
    const history = createMemoryHistory();
    render(
    <Router history={history}>
        <AuthContext.Provider value={{userId: 12}}>
            <ListEachPost 
                id={10}
                text={"a".repeat(140)}
                user={12}
                time={"4 hours ago"}
                username={"user1"}
                engagementRate={10}
            ></ListEachPost>
        </AuthContext.Provider>
    </Router>)
    expect(screen.getByRole("link", {name: /edit this post/i})).toBeInTheDocument()
    expect(screen.getByRole("link", {name: /edit/i})).toHaveAttribute("href", "/home/post/10/update")
    expect(screen.getByRole("link", {name: /see more/i})).toHaveAttribute("href", "/home/post/10")
    expect(screen.getByText("a".repeat(120) + "...")).toBeInTheDocument()
})

test("all posts of other users are rendered properly", () => {
    // users who are not authors shouldn't be able to see "Edit post" link
    const history = createMemoryHistory();
    render(
    <Router history={history}>
        <AuthContext.Provider value={{userId: 2}}>
            <ListEachPost 
                id={1}
                text={"a".repeat(110)}
                user={1}
                time={"4 hours ago"}
                username={"user1"}
                engagementRate={5}
            ></ListEachPost>
        </AuthContext.Provider>
    </Router>)
    expect(screen.getByText("a".repeat(110))).toBeInTheDocument()
    expect(screen.queryByText("a".repeat(110) + "...")).not.toBeInTheDocument()
    expect(screen.getByRole("link", {name: /see more/i})).toHaveAttribute("href", "/home/post/1")
    expect(screen.queryByRole("link", {name: /edit/i})).not.toBeInTheDocument()
    expect(screen.getByRole("link", {name: /see more about this post on another page/i})).toBeInTheDocument()
})