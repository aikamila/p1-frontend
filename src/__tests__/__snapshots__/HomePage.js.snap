// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`all elements are rendered 1`] = `
<div
  className="App"
>
  <header
    className="home-header__background"
    role="banner"
  >
    <div>
      <img
        alt="shangify brand logo"
        className="home-header__img1"
        src="shangify_logo.png"
      />
      <img
        alt=""
        className="home-header__img2"
        src="shangify_brand_name.png"
      />
    </div>
    <nav
      aria-label="main navigation"
      className="home-header__nav"
      role="navigation"
    >
      <a
        aria-label="add a post"
        href="/home/post/add"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="Add a post"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4z"
          />
          <path
            d="M12 2C6.486 2 2 6.486 2 12s4.486 10 10 10 10-4.486 10-10S17.514 2 12 2zm0 18c-4.411 0-8-3.589-8-8s3.589-8 8-8 8 3.589 8 8-3.589 8-8 8z"
          />
        </svg>
      </a>
      <a
        aria-label="homepage"
        href="/home"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="Homepage"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5 22h14a2 2 0 0 0 2-2v-9a1 1 0 0 0-.29-.71l-8-8a1 1 0 0 0-1.41 0l-8 8A1 1 0 0 0 3 11v9a2 2 0 0 0 2 2zm5-2v-5h4v5zm-5-8.59 7-7 7 7V20h-3v-5a2 2 0 0 0-2-2h-4a2 2 0 0 0-2 2v5H5z"
          />
        </svg>
      </a>
      <a
        aria-label="my account"
        href="/home/account/undefined"
        onClick={[Function]}
        style={
          Object {
            "pointerEvents": "auto",
          }
        }
        title="My account"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M3 14s-1 0-1-1 1-4 6-4 6 3 6 4-1 1-1 1H3zm5-6a3 3 0 1 0 0-6 3 3 0 0 0 0 6z"
          />
        </svg>
      </a>
      <button
        disabled={false}
      >
        Log out
      </button>
    </nav>
  </header>
  <main
    className="list-posts__background"
  >
    <div
      className="list-posts__center"
    >
      <div
        className="list-posts__loading"
      >
        <svg
          aria-label="wait a second"
          className="list-posts__loading_sign --spinner"
          fill="none"
          height="1em"
          role="img"
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="2"
            y2="6"
          />
          <line
            x1="12"
            x2="12"
            y1="18"
            y2="22"
          />
          <line
            x1="4.93"
            x2="7.76"
            y1="4.93"
            y2="7.76"
          />
          <line
            x1="16.24"
            x2="19.07"
            y1="16.24"
            y2="19.07"
          />
          <line
            x1="2"
            x2="6"
            y1="12"
            y2="12"
          />
          <line
            x1="18"
            x2="22"
            y1="12"
            y2="12"
          />
          <line
            x1="4.93"
            x2="7.76"
            y1="19.07"
            y2="16.24"
          />
          <line
            x1="16.24"
            x2="19.07"
            y1="7.76"
            y2="4.93"
          />
        </svg>
      </div>
    </div>
  </main>
</div>
`;
